Bump LLVM to v17

mofeing opened the pull request.

Commit:
Update LLVM to v17

Commit:
Update checksums for `clang`, `lld`, `llvm`

Commit:
Fix path to `Triple.h` header

Commit:
Replace count[Leading/Trailing][Zeros/Ones] functions for new methods

Commit:
Replace `Module.getGlobalList().push_back(...)` for `Module.insertGlobalVariable(...)`

Commit:
Replace `Module.getAliasList().push_back(...)` for `Module.insertAlias(...)`

Commit:
Fix access to operand of `MCInstrDesc`

Commit:
Replace `JITEvaluatedSymbol` with  `ExecutorSymbolDef`

Commit:
Fix namespace of `ExecutorSymbolDef`

Commit:
Refactor return type of `findSymbol` and `findUnmangledSymbol`

Commit:
Fix pointer dereference of `Expected<ExecutorSymbolDef>` class

Commit:
Fix `cantFail` call

inkydragon commented:
We've just upgrade to LLVM 16, 1 or 2 weeks ago (#51720). 
That was a long trip, so good luck.

Welcome to the Julia community.

Commit:
Fix deprecated header `llvm/Support/Host.h`

Commit:
Fix `getWithSamePointeeType` deprecation

Commit:
Fix `isOpaquePointer` deprecation

Commit:
Move value on `cantFail` with `Expected`

Commit:
Convert `ExecutorAddr` to `uint64_t` on `JuliaOJIT` addresses

Commit:
Remove deprecated `isOpaqueOrPointeeTypeMatches` calls

Commit:
Remove deprecated `PointerType.isOpaque` calls

Commit:
Link to `libLLVMTargetParser`

Commit:
Fix `StandardInstrumentations.registerCallbacks` call

mofeing commented:
> We've just upgrade to LLVM 16, 1 or 2 weeks ago (#51720). That was a long trip, so good luck.

Yes, I know. I was invited by @gbaraldi and @giordano to try this since getting `LLVM_full_jll` 17 to work was super fast.

> Welcome to the Julia community.

Thanks! :)

mofeing commented:
@gbaraldi would you mind checking 66c13ace24043e3e88b0e9c5f1cd3f5ea21a5cfb?

mofeing commented:
I just found out https://github.com/llvm/llvm-project/blob/23b233c8adad5b81e185e50d04356fab64c2f870/llvm/docs/OpaquePointers.rst#migration-instructions

The following commits might not be completely correct:
- 328ca1a28a45b26725214d8ac33d6127ee7d6867
- 439f3f155dc2c6ced42bdf7813226b9f50cac142
- 8f31fba29b06b91d573771db16882bc5cc1b0a47
- 43c10291fe7aedca65b8917bde4eab037b489f0d

gbaraldi commented:
https://github.com/JuliaLang/julia/commit/66c13ace24043e3e88b0e9c5f1cd3f5ea21a5cfb seems fine, it seems they've changed the order these things need to happen, you can maybe even move the `FAM` down since it's no longer an argument so it's together with the other analysis managers

Commit:
Format `NewPM::run`

Commit:
Fix deprecated `hasSamaElementTypeAs` calls

`PointerType` no longer has an element type, so it always evaluates to true.

Commit:
Fix `supportsTypedPointers`, `getNonOpaquePointerElementType` deprecations

Commit:
Fix deprecated `setOpaquePointers` calls

Opaque pointers are always enabled now.

Commit:
Refactor dead `GetElementPtr` instruction in `AddrspaceRemoveValueMaterializer`

Commit:
Fix unused variables warning but used by `assert`

Commit:
Remove dead code

giordano commented on src/jitlayers.cpp:
```diff
@@ -1665,17 +1659,6 @@ void fixupTM(TargetMachine &TM) {
-void SetOpaquePointer(LLVMContext &ctx) {
-    if (jl_opaque_ptrs_set)
-        return;
-#ifndef JL_LLVM_OPAQUE_POINTERS
```
If opaque pointers are always supposed to exist, then the definition of `JL_LLVM_OPAQUE_POINTERS` can be removed from https://github.com/JuliaLang/julia/blob/d298c44bc60eebdfc77b7f9c64ca913af1e5079d/src/llvm-version.h#L17-L19 this seems to be the only place where [the macro was used](https://github.com/search?q=repo%3AJuliaLang/julia%20JL_LLVM_OPAQUE_POINTERS&type=code).

giordano commented:
```diff
diff --git a/src/intrinsics.cpp b/src/intrinsics.cpp
index c784727c4f5..ff3c55bc072 100644
--- a/src/intrinsics.cpp
+++ b/src/intrinsics.cpp
@@ -4,6 +4,10 @@ namespace JL_I {
 #include "intrinsics.h"
 }
 
+#include <array>
+#include <bitset>
+#include <string>
+
 #include "ccall.cpp"
 
 //Mark our stats as being from intrinsics irgen
```
fixes compilation of `src/codegen.cpp` for me on gnu linux.

Commit:
Remove definition of `JL_LLVM_OPAQUE_POINTERS`

Commit:
Fix missing stdlib headers

Co-authored-by: Mosè Giordano <mose@gnu.org>

Commit:
Add missing checksums

giordano committed:
Try to fix checksums

I used the command
```
make -f contrib/refresh_checksums.mk -j2 clang llvm
```
with `-j2` instead of `-j` to try and reduce race conditions.

Commit:
Fix LLVM checksums

giordano committed:
Fix Clang checksums

```
rm -rf deps/srccache/Clang*
make -f contrib/refresh_checksums.mk -j4 clang
```

giordano committed:
[libLLVM_jll] Fix test

Function `LLVMInitializeTarget` was removed from C API of libLLVM in version 17.
Let's use a different function, the goal of the test is a simple sanity check,
we don't have to do anything with that function.

giordano committed:
Fix LLVM checksums take 3

```
rm -rf deps/srccache/*LLVM*
time make -f contrib/refresh_checksums.mk -j4 llvm
```

Commit:
Fix `ExecutorSymbolDef` evaluation to `bool`

Commit:
Fix assertion

vchuravy commented:
@mofeing thank you for your work! One note from my side is that we aim to support the lastest LLVM version used by the latest Julia release for 1.10 that would be LLVM15. The goal is to be able to build for LLVM 15 without too many issues to be able to check where regressions came from. 

This means that you may need to guard your changes under a `JL_LLVM_VERSION >= 170000`

Commit:
Make code compatible with LLVM 15 again

mofeing commented on stdlib/libLLVM_jll/test/runtests.jl:
```diff
@@ -3,5 +3,6 @@
 
 @testset "libLLVM_jll" begin
-    @test dlsym(libLLVM_jll.libLLVM_handle, :LLVMInitializeTarget; throw_error=false) !== nothing
+    # Try to find a symbol from the C API of libLLVM as a simple sanity check.
+    @test dlsym(libLLVM_jll.libLLVM_handle, :LLVMContextCreate; throw_error=false) !== nothing
```
@vchuravy Not sure how to make this code compatible with LLVM 15 back again. Is there some way to check the version of `libLLVM_jll`?

giordano replied:
Isn't `LLVMContextCreate` available in LLVM 15?  When I [blamed that function](https://github.com/llvm/llvm-project/blame/6009708b4367171ccdbf4b5905cb6a803753fe18/llvm/include/llvm-c/Core.h#L518) I found it was introduced 15 years ago.

mofeing commented:
> @mofeing thank you for your work! One note from my side is that we aim to support the lastest LLVM version used by the latest Julia release for 1.10 that would be LLVM15. The goal is to be able to build for LLVM 15 without too many issues to be able to check where regressions came from.
> 
> This means that you may need to guard your changes under a `JL_LLVM_VERSION >= 170000`

Understood! I've refactored the code to include some `JL_LLVM_VERSION` guards.

giordano commented on src/Makefile:
```diff
@@ -79,7 +79,7 @@ endif
 RT_LLVM_LIBS := support
 
-ifeq ($(LLVM_VER_MAJ),16)
+ifeq ($(shell test $(LLVM_VER_MAJ) -ge 16),1)
```
On this branch I'm getting
```console
% make -C src print-RT_LLVM_LIBS
RT_LLVM_LIBS=support
```
I don't think this is working.  Maybe do https://github.com/JuliaLang/julia/pull/53083#discussion_r1469963998?

mofeing replied:
I managed to fix it by adding `&& echo true`.

giordano replied:
Note that #53083 was merged and that was working for me with both llvm 16 and 17. You now need to rebase on master because of the conflicts.

mofeing replied:
Ah damm. Okay.

giordano commented on src/aotcompile.cpp:
```diff
@@ -1649,7 +1659,9 @@ void jl_dump_native_impl(void *native_code,
         LLVMContext Context;
-        SetOpaquePointer(Context);
+        #if JL_LLVM_VERSION < 170000
+        SetOpaquePointer(ctx);
```
Should this be
```suggestion
        SetOpaquePointer(Context);
```
?  Same below.

giordano commented on src/runtime_intrinsics.c:
```diff
@@ -1566,6 +1566,16 @@ bi_iintrinsic_cnvtb_fast(LLVMAShr, ashr_op, ashr_int, , 1)
 //un_iintrinsic_fast(LLVMByteSwap, bswap_op, bswap_int, u)
 un_iintrinsic_slow(LLVMByteSwap, bswap_int, u)
 //#define ctpop_op(a) __builtin_ctpop(a)
+#if JL_LLVM_VERSION >= 170000
```
Need to include
```c
#include "llvm-version.h"
```
at the top of this file to define the macro.

mofeing committed:
Fix code

Co-authored-by: Mosè Giordano <giordano@users.noreply.github.com>

Commit:
Fix Makefile conditional linking

Commit:
Fix header inclusion

Commit:
Fix LLVM version requirement for `libLLVMTargetParser` link

Commit:
Fix typo

Commit:
Fix header inclusion

Commit:
Fix C++ header on C code

Commit:
Fix header inclusion after rebase

Commit:
Try fix for `is_trivially_copy/move_constructible`

Commit:
Add missing header

mofeing commented:
I managed to fix the build for `i686`! There are just 2 build errors remaining:
- This one appears in assert builds (`x86_64-linux-gnuassert` and `powerpc64le-linux-gnuassert`) and `llvmpasses`:
```make
julia: /workspace/srcdir/llvm-project/llvm/lib/IR/Constants.cpp:1484: llvm::Constant* llvm::ConstantExpr::getWithOperands(llvm::ArrayRef<llvm::Constant*>, llvm::Type*, bool, llvm::Type*) const: Assertion `SrcTy || (Ops[0]->getType() == getOperand(0)->getType())' failed.
[11094] signal 6 (-6): Aborted
in expression starting at none:0
Allocations: 23086645 (Pool: 23086341; Big: 304); GC: 25
Aborted
*** This error is usually fixed by running `make clean`. If the error persists, try `make cleanall`. ***
make[1]: *** [sysimage.mk:96: /cache/build/builder-amdci4-3/julialang/julia-master/usr/lib/julia/sys-o.a] Error 1
make: *** [Makefile:109: julia-sysimg-release] Error 2
```
- This one appears in `asan`:
```make
ERROR: Unable to load dependent library /cache/build/default-jelaqua-0/julialang/julia-master/tmp/test-asan/asan/usr/bin/../lib/libjulia-internal-debug.so.1.11
Message:/lib/x86_64-linux-gnu/libgcc_s.so.1: version `GCC_13.0.0' not found (required by /cache/build/default-jelaqua-0/julialang/julia-master/tmp/test-asan/asan/usr/bin/../lib/libjulia-internal-debug.so.1.11)
```

Also, `aarch64-apple-darwin` is failing on test with a big segmentation fault when testing `ccall`:
```make
Executing tests that run on node 1 only:
ccall                                            (1) |        started at 2024-01-30T06:16:59.068
[95758] signal 11 (2): Segmentation fault: 11
in expression starting at /Users/julia/.julia/scratchspaces/a66863c6-20e8-4ff4-8a62-49f30b1f605e/agent-cache/default-honeycrisp-XC9YQX9HH2.0/build/default-honeycrisp-XC9YQX9HH2-0/julialang/julia-master/julia-63db41f1c5/share/julia/test/ccall.jl:785
_ZN4llvm13MemCpyOptPass13processMemCpyEPNS_10MemCpyInstERNS_14ilist_iteratorINS_12ilist_detail12node_optionsINS_11InstructionELb0ELb0EvEELb0ELb0EEE at /Users/julia/.julia/scratchspaces/a66863c6-20e8-4ff4-8a62-49f30b1f605e/agent-cache/default-honeycrisp-XC9YQX9HH2.0/build/default-honeycrisp-XC9YQX9HH2-0/julialang/julia-master/julia-63db41f1c5/lib/julia/libLLVM.dylib (unknown line)
```

gbaraldi commented:
The assertion failure is probably the most pressing.

mofeing commented on src/cgutils.cpp:
```diff
@@ -3282,7 +3296,11 @@ static void recursively_adjust_ptr_type(llvm::Value *Val, unsigned FromAS, unsig
         if (isa<GetElementPtrInst>(User)) {
             GetElementPtrInst *Inst = cast<GetElementPtrInst>(User);
+            #if JL_LLVM_VERSION >= 170000
+            Inst->mutateType(PointerType::get(Inst->getType(), ToAS));
```
@gbaraldi I believe this is the root of the assertion crash. The erroring assert is in https://github.com/llvm/llvm-project/blob/6009708b4367171ccdbf4b5905cb6a803753fe18/llvm/lib/IR/Constants.cpp#L1484, which processes a `GetElementPtr` instruction.

Commit:
Try fix assertion crash in `ConstantExpr::getWithOperands`

Commit:
Replace LLVM version guard for comment

Commit:
Remove legacy `-enable-new-pm` flag from llvmpasses

Commit:
Fix undefined identifier on LLVM 17

Commit:
Revert "Fix undefined identifier on LLVM 17"

This reverts commit 414db35380de004c14c7385c41006bdce1d18ea0.

Commit:
Revert "Replace LLVM version guard for comment"

This reverts commit 14712d8026d94aa81c0966f755b4cd294ff9d2f8.

Commit:
Remove `--opaque-pointers` flags on llvmpasses

Commit:
Fix missing `RUN` statement

giordano commented:
For the [issue during ASAN job](https://buildkite.com/julialang/julia-master/builds/32998#018d648a-e4a3-4fd2-a3d0-29cdb3f56dee/712-1319):
```
    JULIA tmp/test-asan/asan/usr/lib/julia/corecompiler.ji
ERROR: Unable to load dependent library /cache/build/builder-amdci4-6/julialang/julia-master/tmp/test-asan/asan/usr/bin/../lib/libjulia-internal-debug.so.1.11
Message:/lib/x86_64-linux-gnu/libgcc_s.so.1: version `GCC_13.0.0' not found (required by /cache/build/builder-amdci4-6/julialang/julia-master/tmp/test-asan/asan/usr/bin/../lib/libjulia-internal-debug.so.1.11)
make[1]: *** [/cache/build/builder-amdci4-6/julialang/julia-master/sysimage.mk:64: /cache/build/builder-amdci4-6/julialang/julia-master/tmp/test-asan/asan/usr/lib/julia/corecompiler.ji] Error 1
make[1]: Leaving directory '/cache/build/builder-amdci4-6/julialang/julia-master/tmp/test-asan/asan'
make: *** [/cache/build/builder-amdci4-6/julialang/julia-master/Makefile:103: julia-sysimg-ji] Error 2
make: Leaving directory '/cache/build/builder-amdci4-6/julialang/julia-master/tmp/test-asan/asan'
```
Feels like something is missing RPATH settings.  For reference, look at the [`x86_64-linux-gnu` build](https://buildkite.com/julialang/julia-master/builds/32998#018d648a-d859-47d6-952b-e65a71052170/744-1320):
```
 gcc -m64 -std=gnu11 -pipe -fPIC -fno-strict-aliasing -D_FILE_OFFSET_BITS=64 -fno-gnu-unique -I/cache/build/tester-amdci5-9/julialang/julia-master/src -I/cache/build/tester-amdci5-9/julialang/julia-master/src -I/cache/build/tester-amdci5-9/julialang/julia-master/src/support -I/cache/build/tester-amdci5-9/julialang/julia-master/usr/include -ffreestanding -DGLIBCXX_LEAST_VERSION_SYMBOL=\"GLIBCXX_3.4.33\" -O3 -ggdb2 -falign-functions -momit-leaf-frame-pointer -DDEP_LIBS="\"libgcc_s.so.1:libopenlibm.so:@libstdc++.so.6:@libjulia-internal.so.1.11:@libjulia-codegen.so.1.11:\"" ./loader_exe.o -o /cache/build/tester-amdci5-9/julialang/julia-master/usr/bin/julia -Wl,-Bdynamic -Wl,-no-undefined -ffreestanding -L/cache/build/tester-amdci5-9/julialang/julia-master/usr/lib -L/cache/build/tester-amdci5-9/julialang/julia-master/usr/lib -Wl,--no-as-needed -ldl -lpthread -rdynamic -lc -Wl,--as-needed -Wl,-z,notext -Wl,-rpath,'$ORIGIN/../lib' -Wl,-rpath,'$ORIGIN/../lib/julia' -Wl,-rpath-link,/cache/build/tester-amdci5-9/julialang/julia-master/usr/lib -Wl,-z,origin -Wl,--enable-new-dtags -ljulia
```
When the `julia` executable is linked, the rpath is set correctly.  But I don't know what's going on in the ASAN build, because that's not built in verbose mode.  That's an issue in the CI setup.

mofeing committed:
Merge branch 'master' into bump-llvm17

gbaraldi committed:
Merge branch 'master' into bump-llvm17

mofeing committed:
Merge branch 'master' into bump-llvm17

mofeing commented:
@gbaraldi the segmentation fault in `ccall` test seems to come from `llvm::MemCpyOptPass::processMemCpy`.
Maybe related to https://github.com/llvm/llvm-project/issues/71183?

giordano commented:
> or should we cherry-pick ourselves in Julia's LLVM fork?

I don't know what's LLVM policy for backporting fixes, but we definitely usually backport them in https://github.com/JuliaLang/llvm-project if important for Julia.

gbaraldi commented:
Oh that didn't get in? We have that patch in LLVM 16 so we need to add it to our fork of 17.

mofeing commented:
Just backported https://github.com/llvm/llvm-project/commit/5c3beb7b1e26d38b0933a28432dfbce4e00cf329, tested in local and it works! All tests pass!

```
Executing tests that run on node 1 only:
ccall                                            (1) |        started at 2024-03-04T17:44:02.738
ccall                                            (1) |    18.43 |   0.16 |  0.8 |     547.49 |   480.39
precompile                                       (1) |        started at 2024-03-04T17:44:21.173

...

Test Summary: |     Pass  Broken     Total      Time
  Overall     | 69374577    3219  69377796  27m43.1s
    SUCCESS
```

Will relaunch build when https://github.com/JuliaLang/llvm-project/pull/25 is merged and `LLVM_jll` is updated accordingly.

Now the only problem is the ASAN job 😜

Commit:
Refresh checksums

mofeing commented:
Now everything builds! Only thing remaining is the failing ASAN job @gbaraldi @giordano 

Btw, about Tier 3 platforms:
- `x86_64-linux-musl` builds successfully but crashes into a OOM error when generating the sysimage https://buildkite.com/julialang/julia-master/builds/34384#018e1366-9fdd-4b6d-b095-587646daa398/742-2470
- `aarch64-linux-gnu` fails on test job due to a unexpected EOF https://buildkite.com/julialang/julia-master/builds/34384#018e1366-94ca-4991-bf04-1b364572e596/809-973
  - `powerpcle64-linux-gnu` seems to have several of this EOF errors

giordano commented:
Those are expected:

* musl is #52707
* aarch64-linux-gnu is #52434

I have no idea what's the problem with the asan build and I couldn't get any insights from https://github.com/JuliaCI/julia-buildkite/pull/339.  @staticfloat would you be able to have a look at https://github.com/JuliaLang/julia/pull/53070#issuecomment-1921241959?

giordano commented:
@mofeing at https://github.com/JuliaLang/julia/blob/5023ee21d70b734edf206aab3cac7c202ee0235a/sysimage.mk#L65 can you export `LD_DEBUG=libs` to try and get more information about why the julia process is failing to load the dependencies?  If that's not enough you could try `LD_DEBUG=all`, but the log file will probably be crazy large, but [sometimes it's more useful](https://github.com/JuliaPackaging/Yggdrasil/pull/5028#issuecomment-1166377616) than `libs` alone.

Commit:
Log debug info from lib loading

giordano commented:
https://buildkite.com/julialang/julia-master/builds/34691#018e3d3a-df06-43e0-8657-e0676f0129d7/723-1505
```
     17129:	/lib/x86_64-linux-gnu/libstdc++.so.6: error: symbol lookup error: undefined symbol: GLIBCXX_3.4.33 (fatal)
     17128:	find library=libjulia-internal-debug.so.1.12 [0]; searching
     17128:	 search cache=/etc/ld.so.cache
     17128:	 search path=/lib/x86_64-linux-gnu:/usr/lib/x86_64-linux-gnu:/lib:/usr/lib		(system search path)
     17128:	  trying file=/lib/x86_64-linux-gnu/libjulia-internal-debug.so.1.12
     17128:	  trying file=/usr/lib/x86_64-linux-gnu/libjulia-internal-debug.so.1.12
     17128:	  trying file=/lib/libjulia-internal-debug.so.1.12
     17128:	  trying file=/usr/lib/libjulia-internal-debug.so.1.12
```
So this confirms that `libjulia-internal-debug.so.1.12` is trying to be loaded without using rpath, but in https://github.com/JuliaCI/julia-buildkite/pull/339#issuecomment-1980766664 it looked like rpath should be set.  I still don't understand what's going on.  @staticfloat would you be able to have a look at this?  This is the last error we're facing here.

vtjnash commented:
Is that this bug: https://github.com/JuliaPackaging/Yggdrasil/issues/3703 or did the rpath of LLVM ever get changed to be set correctly and that issue is stale?

giordano commented:
Locally, for a release build, I get
```console
% for x in usr/lib/*.so; do echo $x; objdump -x $x | grep 'RPATH\|RUNPATH'; done
usr/lib/libamd.so
  RPATH                $ORIGIN
usr/lib/libasan.so
  RUNPATH              $ORIGIN
usr/lib/libatomic.so
usr/lib/libblastrampoline.so
usr/lib/libbtf.so
  RPATH                $ORIGIN
usr/lib/libcamd.so
  RPATH                $ORIGIN
usr/lib/libccalllazybar.so
usr/lib/libccalllazyfoo.so
usr/lib/libccalltest.so
usr/lib/libccolamd.so
  RPATH                $ORIGIN
usr/lib/libcholmod.so
  RPATH                $ORIGIN
usr/lib/libcolamd.so
  RPATH                $ORIGIN
usr/lib/libcurl.so
  RUNPATH              $ORIGIN
usr/lib/libdSFMT.so
usr/lib/libgfortran.so
  RUNPATH              $ORIGIN
usr/lib/libgit2.so
  RPATH                $ORIGIN
usr/lib/libgmp.so
usr/lib/libgmpxx.so
  RUNPATH              $ORIGIN
usr/lib/libgomp.so
usr/lib/libhwasan.so
  RUNPATH              $ORIGIN
usr/lib/libitm.so
usr/lib/libjulia-codegen.so
  RUNPATH              $ORIGIN
usr/lib/libjulia-internal.so
  RUNPATH              $ORIGIN
usr/lib/libjulia.so
usr/lib/libklu_cholmod.so
  RPATH                $ORIGIN
usr/lib/libklu.so
  RPATH                $ORIGIN
usr/lib/libldl.so
  RPATH                $ORIGIN
usr/lib/libLLVM-16.0.6jl.so
  RPATH                $ORIGIN/../lib
usr/lib/libLLVM-16jl.so
  RPATH                $ORIGIN/../lib
usr/lib/libllvmcalltest.so
usr/lib/libLLVM.so
  RPATH                $ORIGIN/../lib
usr/lib/liblsan.so
  RUNPATH              $ORIGIN
usr/lib/libmbedcrypto.so
usr/lib/libmbedtls.so
  RUNPATH              $ORIGIN
usr/lib/libmbedx509.so
  RUNPATH              $ORIGIN
usr/lib/libmpfr.so
  RUNPATH              $ORIGIN
usr/lib/libnghttp2.so
usr/lib/libobjc.so
  RUNPATH              $ORIGIN
usr/lib/libopenblas64_.0.3.26.so
  RPATH                $ORIGIN
usr/lib/libopenblas64_.so
  RPATH                $ORIGIN
usr/lib/libopenlibm.so
usr/lib/libpcre2-16.so
usr/lib/libpcre2-32.so
usr/lib/libpcre2-8.so
usr/lib/libpcre2-posix.so
  RUNPATH              $ORIGIN
usr/lib/libquadmath.so
usr/lib/librbio.so
  RPATH                $ORIGIN
usr/lib/libspqr.so
  RPATH                $ORIGIN
usr/lib/libssh2.so
  RUNPATH              $ORIGIN
usr/lib/libssp.so
usr/lib/libstdc++.so
  RUNPATH              $ORIGIN
usr/lib/libsuitesparseconfig.so
  RPATH                $ORIGIN
usr/lib/libtsan.so
  RUNPATH              $ORIGIN
usr/lib/libubsan.so
  RUNPATH              $ORIGIN
usr/lib/libumfpack.so
  RPATH                $ORIGIN
usr/lib/libunwind-coredump.so
  RUNPATH              $ORIGIN
usr/lib/libunwind-generic.so
  RUNPATH              $ORIGIN
usr/lib/libunwind-ptrace.so
  RUNPATH              $ORIGIN
usr/lib/libunwind-setjmp.so
  RUNPATH              $ORIGIN
usr/lib/libunwind.so
  RUNPATH              $ORIGIN
usr/lib/libunwind-x86_64.so
  RUNPATH              $ORIGIN
usr/lib/libuv.so
usr/lib/libz.so
```
Issue https://github.com/JuliaPackaging/Yggdrasil/issues/3703 may have been fixed since?

yuyichao commented:
A normal build of the debug version seems to fail on linux with this due to linking issue with `__stack_chk_guard`. This seems to be caused by https://github.com/llvm/llvm-project/commit/e018cbf7208b3d34f18997ddee84c66cee32fb1b which no longer gets the relocation type from the target machine but the module flags instead.

Setting the PIC flag on the module by adding `dataM.setPICLevel(PICLevel::BigPIC);` [here](https://github.com/JuliaLang/julia/blob/1d532f4fc798ecaa5bcacfcedc4c36143ad0e68f/src/aotcompile.cpp#L1692) seems to work.

gbaraldi committed:
Merge branch 'master' into bump-llvm17

gbaraldi committed:
Merge branch 'master' into bump-llvm17

gbaraldi committed:
Add runpath to libjulia

gbaraldi committed:
Patch libclang to have the correct rpath

gbaraldi committed:
Do it in the right place

gbaraldi committed:
Simplify

gbaraldi committed:
Try something different

gbaraldi committed:
Install patchelf

gbaraldi committed:
Moar dollars

gbaraldi committed:
Even more dollars

gbaraldi committed:
Add patchelf override to the makefile

gbaraldi committed:
Add BigPIC

gbaraldi commented:
@nanosoldier `runbenchmarks(vs=":master")`

gbaraldi commented:
@nanosoldier `runbenchmarks(ALL, vs=":master")`

gbaraldi commented:
@nanosoldier `runtests()`

giordano reviewed (commented):
@gbaraldi can you please move the changes in https://github.com/JuliaLang/julia/compare/b1d87b9374b9adeea9802d6fe14853d5ad7e8119...86f142901094b0c8891b373121c9dbbdbb8b1618 to a separate PR and explain the rationale.  These are for addressing https://github.com/JuliaLang/julia/pull/53070#issuecomment-1921241959, but they are logically separate from this PR, they're fixing a bug in the build system, it's just an accident that these are now needed with llvm 17.

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/53070#issuecomment-2057786953) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/6560910_vs_c0611e8/report.md).

nanosoldier commented:
The package evaluation job [you requested](https://github.com/JuliaLang/julia/pull/53070#issuecomment-2057844491) has completed - possible new issues were detected.
The [**full report**](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/13a077a_vs_f085913/report.html) is available.

gbaraldi committed:
Merge branch 'master' into bump-llvm17

gbaraldi committed:
Merge branch 'master' into bump-llvm17

giordano committed:
Apply suggestions from code review

mofeing committed:
Merge branch 'master' into bump-llvm17

Zentrik commented:
```
diff --git a/src/llvm-alloc-opt.cpp b/src/llvm-alloc-opt.cpp
index b08e18068a..25c686827b 100644
--- a/src/llvm-alloc-opt.cpp
+++ b/src/llvm-alloc-opt.cpp
@@ -1076,7 +1076,7 @@ void Optimizer::splitOnStack(CallInst *orig_inst)
                     store_ty = T_pjlvalue;
                 }
                 else {
-                    store_ty = PointerType::getWithSamePointeeType(T_pjlvalue, store_ty->getPointerAddressSpace());
+                    store_ty = PointerType::get(T_pjlvalue->getContext(), store_ty->getPointerAddressSpace());
                     store_val = builder.CreateBitCast(store_val, store_ty);
                 }
                 if (store_ty->getPointerAddressSpace() != AddressSpace::Tracked)
```
should fix the build error.

giordano commented:
Can you make a change suggestion?

mofeing commented:
Sorry, my fault as I put it as in https://github.com/JuliaLang/julia/commit/074ea2a47f99e21191e262b89aa3aefea87a18d6#diff-df780eefd7ac77d62e69323896eeb93d627bd18e48a8e48a4231e379e0c86efeR1070 but `PointerType::getWithSamePointeeType` was removed.

Can fix it in a moment

Zentrik commented on src/llvm-alloc-opt.cpp:
```diff
@@ -1067,8 +1076,7 @@ void Optimizer::splitOnStack(CallInst *orig_inst)
                 else {
-                    store_ty = PointerType::getWithSamePointeeType(
-                        T_pjlvalue, store_ty->getPointerAddressSpace());
+                    store_ty = PointerType::getWithSamePointeeType(T_pjlvalue, store_ty->getPointerAddressSpace());
```
```suggestion
                    store_ty = PointerType::get(T_pjlvalue->getContext(), store_ty->getPointerAddressSpace());
```

mofeing committed:
Fix call to removed `getWithSamePointeeType`

Co-authored-by: Zentrik <Zentrik@users.noreply.github.com>

mofeing committed:
Merge branch 'master' into bump-llvm17

mofeing commented on src/llvm-alloc-opt.cpp:
```diff
@@ -1067,8 +1076,7 @@ void Optimizer::splitOnStack(CallInst *orig_inst)
                 else {
-                    store_ty = PointerType::getWithSamePointeeType(
-                        T_pjlvalue, store_ty->getPointerAddressSpace());
+                    store_ty = PointerType::get(T_pjlvalue->getContext(), store_ty->getPointerAddressSpace());
```
@maleadt We refactored this line to merge the code from #54113 but now the `gc-invariant-verifier` test is failing https://buildkite.com/julialang/julia-master/builds/35738#018f0295-5279-4b99-818d-36905524f086

Would you mind checking it out please?

mofeing replied:
done!

Commit:
Remove legacy flags `enable-new-pm`,`opaque-pointers` in LLVM passes tests

giordano commented:
Before doing a full re-test of all packages (I guess we want to wait for https://github.com/JuliaPackaging/Yggdrasil/pull/8510 for that), I wanted to quickly see if #54113 fixed most of the issues we saw [above](https://github.com/JuliaLang/julia/pull/53070#issuecomment-2059537551)


@nanosoldier `runtests(["AssociatedLegendrePolynomials", "MeanFilters", "StringAlgorithms", "EarlyStopping", "EvilArrays", "OffsetArrays", "MosaicViews", "StrBase", "UniformIsingModels", "GPUCompiler", "StructArrays", "HistogramThresholding", "ParallelUtilities", "ImageShow", "LegendrePolynomials", "ExtractLinks", "ImageMetadata", "MetaImageFormat", "ImageBinarization", "Sixel", "ImageTransformations", "Neurthino", "ImagineFormat", "BenchmarkProfiles", "Pyehtim", "NonlinearSystems", "Permanents", "FMIImport", "NRRD", "PreallocationTools", "TemplateMatching", "TensorBoardLogger", "GIFImages", "LabelledArrays", "TriangularSolve", "ImageFiltering", "TropicalGEMM", "OMETIFF", "GeoArrayOps", "Geomorphometry", "MRICoilSensitivities", "ImageEdgeDetection", "TuringCallbacks", "TestImages", "TimeseriesSurrogates", "MicroscopyLabels", "Jadex", "ThermodynamicIntegration", "FinEtools", "ImageQualityIndexes", "ImageSegmentation", "TinyGibbs", "ReservoirComputing", "JumpProcesses", "BayesQR", "SurveyDataWeighting", "DeepFry", "MCMCChains", "RandomFeatures", "MCMCChainsStorage", "DifferentialEvolutionMCMC", "CombinatorialSpaces", "MCMCTempering", "ImageView", "TidierDB", "MRIReco", "Dolo", "SimSpin", "DistributedStwdLDA", "CLEARSWI", "TidierFiles", "ROMEO", "GeneralizedSDistributions", "MLJTestIntegration", "MriResearchTools", "Images", "Agents", "ImageTracking", "SurfaceReactions", "SurfaceCoverage", "DifferentialEquations", "EqualitySampler", "LongwaveModePropagator", "ImageFeatures", "SafetySignalDetection", "ImageUtils", "ActionModels", "PlugFlowReactor", "BLASBenchmarksCPU", "HetaSimulator", "Ai4EComponentLib", "SDEProblemLibrary", "CellMLToolkit", "ProbabilisticEchoInversion", "LibTrixi", "ClimateTools", "LiquidElectrolytes", "EarthSciMLBase", "ClimatePlots", "Collide", "PowerGraphics", "MRINavigator", "Turkie"])`

gbaraldi commented:
That change should only affect performance, but yeah waiting doesn't hurt.

nanosoldier commented:
The package evaluation job [you requested](https://github.com/JuliaLang/julia/pull/53070#issuecomment-2070687280) has completed - possible new issues were detected.
The [**full report**](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/1b5ee47_vs_903ffff/report.html) is available.

gbaraldi committed:
Update LLVM to 17.0.6+2

gbaraldi committed:
Also update lld checksums

giordano commented:
Good news: #54113 did fix a large chunk of failing packages, now we're down from 103 to 40 failing tests, only 17 of which started failing on this PR.

Most common error now seems to be (e.g. [PreallocationTools.jl](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/1b5ee47_vs_903ffff/PreallocationTools.primary.log))
```
[887] signal 6 (-6): Aborted
in expression starting at /home/pkgeval/.julia/packages/PreallocationTools/JFu7v/test/core_odes.jl:18
unknown function (ip: 0x7fb140cb7d3c)
gsignal at /lib/x86_64-linux-gnu/libc.so.6 (unknown line)
abort at /lib/x86_64-linux-gnu/libc.so.6 (unknown line)
unknown function (ip: 0x7fb140c53394)
__assert_fail at /lib/x86_64-linux-gnu/libc.so.6 (unknown line)
_ZN4llvm15ScalarEvolution15willNotOverflowENS_11Instruction9BinaryOpsEbPKNS_4SCEVES5_PKS1_ at /opt/julia/bin/../lib/julia/libLLVM-17jl.so (unknown line)
_ZN12_GLOBAL__N_119InductiveRangeCheck26extractRangeChecksFromCondEPN4llvm4LoopERNS1_15ScalarEvolutionERNS1_3UseERNS1_15SmallVectorImplIS0_EERNS1_15SmallPtrSetImplIPNS1_5ValueEEE at /opt/julia/bin/../lib/julia/libLLVM-17jl.so (unknown line)
_ZN4llvm8IRCEPass3runERNS_8FunctionERNS_15AnalysisManagerIS1_JEEE at /opt/julia/bin/../lib/julia/libLLVM-17jl.so (unknown line)
run at /source/usr/include/llvm/IR/PassManagerInternal.h:89
run at /source/usr/include/llvm/IR/PassManager.h:517 [inlined]
run at /source/usr/include/llvm/IR/PassManagerInternal.h:89
_ZN4llvm27ModuleToFunctionPassAdaptor3runERNS_6ModuleERNS_15AnalysisManagerIS1_JEEE at /opt/julia/bin/../lib/julia/libLLVM-17jl.so (unknown line)
run at /source/usr/include/llvm/IR/PassManagerInternal.h:89
_ZN4llvm11PassManagerINS_6ModuleENS_15AnalysisManagerIS1_JEEEJEE3runERS1_RS3_ at /opt/julia/bin/../lib/julia/libLLVM-17jl.so (unknown line)
run at /source/src/pipeline.cpp:837
operator() at /source/src/jitlayers.cpp:1244
withModuleDo<(anonymous namespace)::OptimizerT<N>::operator()(llvm::orc::ThreadSafeModule, llvm::orc::MaterializationResponsibility&) [with long unsigned int N = 4]::<lambda(llvm::Module&)> > at /source/usr/include/llvm/ExecutionEngine/Orc/ThreadSafeModule.h:136 [inlined]
operator() at /source/src/jitlayers.cpp:1205 [inlined]
CallImpl<(anonymous namespace)::OptimizerT<4> > at /source/usr/include/llvm/ADT/FunctionExtras.h:220
_ZN4llvm3orc16IRTransformLayer4emitESt10unique_ptrINS0_29MaterializationResponsibilityESt14default_deleteIS3_EENS0_16ThreadSafeModuleE at /opt/julia/bin/../lib/julia/libLLVM-17jl.so (unknown line)
_ZN4llvm3orc16IRTransformLayer4emitESt10unique_ptrINS0_29MaterializationResponsibilityESt14default_deleteIS3_EENS0_16ThreadSafeModuleE at /opt/julia/bin/../lib/julia/libLLVM-17jl.so (unknown line)
_ZN4llvm3orc16IRTransformLayer4emitESt10unique_ptrINS0_29MaterializationResponsibilityESt14default_deleteIS3_EENS0_16ThreadSafeModuleE at /opt/julia/bin/../lib/julia/libLLVM-17jl.so (unknown line)
_ZN4llvm3orc31BasicIRLayerMaterializationUnit11materializeESt10unique_ptrINS0_29MaterializationResponsibilityESt14default_deleteIS3_EE at /opt/julia/bin/../lib/julia/libLLVM-17jl.so (unknown line)
_ZN4llvm3orc19MaterializationTask3runEv at /opt/julia/bin/../lib/julia/libLLVM-17jl.so (unknown line)
_ZN4llvm6detail18UniqueFunctionBaseIvJSt10unique_ptrINS_3orc4TaskESt14default_deleteIS4_EEEE8CallImplIPFvS7_EEEvPvRS7_ at /opt/julia/bin/../lib/julia/libLLVM-17jl.so (unknown line)
_ZN4llvm3orc16ExecutionSession22dispatchOutstandingMUsEv at /opt/julia/bin/../lib/julia/libLLVM-17jl.so (unknown line)
_ZN4llvm3orc16ExecutionSession17OL_completeLookupESt10unique_ptrINS0_21InProgressLookupStateESt14default_deleteIS3_EESt10shared_ptrINS0_23AsynchronousSymbolQueryEESt8functionIFvRKNS_8DenseMapIPNS0_8JITDylibENS_8DenseSetINS0_15SymbolStringPtrENS_12DenseMapInfoISF_vEEEENSG_ISD_vEENS_6detail12DenseMapPairISD_SI_EEEEEE at /opt/julia/bin/../lib/julia/libLLVM-17jl.so (unknown line)
_ZN4llvm3orc25InProgressFullLookupState8completeESt10unique_ptrINS0_21InProgressLookupStateESt14default_deleteIS3_EE at /opt/julia/bin/../lib/julia/libLLVM-17jl.so (unknown line)
_ZN4llvm3orc16ExecutionSession19OL_applyQueryPhase1ESt10unique_ptrINS0_21InProgressLookupStateESt14default_deleteIS3_EENS_5ErrorE at /opt/julia/bin/../lib/julia/libLLVM-17jl.so (unknown line)
_ZN4llvm3orc16ExecutionSession6lookupENS0_10LookupKindERKSt6vectorISt4pairIPNS0_8JITDylibENS0_19JITDylibLookupFlagsEESaIS8_EENS0_15SymbolLookupSetENS0_11SymbolStateENS_15unique_functionIFvNS_8ExpectedINS_8DenseMapINS0_15SymbolStringPtrENS0_17ExecutorSymbolDefENS_12DenseMapInfoISI_vEENS_6detail12DenseMapPairISI_SJ_EEEEEEEEESt8functionIFvRKNSH_IS6_NS_8DenseSetISI_SL_EENSK_IS6_vEENSN_IS6_SV_EEEEEE at /opt/julia/bin/../lib/julia/libLLVM-17jl.so (unknown line)
_ZN4llvm3orc16ExecutionSession6lookupERKSt6vectorISt4pairIPNS0_8JITDylibENS0_19JITDylibLookupFlagsEESaIS7_EENS0_15SymbolLookupSetENS0_10LookupKindENS0_11SymbolStateESt8functionIFvRKNS_8DenseMapIS5_NS_8DenseSetINS0_15SymbolStringPtrENS_12DenseMapInfoISI_vEEEENSJ_IS5_vEENS_6detail12DenseMapPairIS5_SL_EEEEEE at /opt/julia/bin/../lib/julia/libLLVM-17jl.so (unknown line)
addModule at /source/src/jitlayers.cpp:1868
jl_add_to_ee at /source/src/jitlayers.cpp:2359
jl_add_to_ee at /source/src/jitlayers.cpp:2338
jl_add_to_ee at /source/src/jitlayers.cpp:2338
jl_add_to_ee at /source/src/jitlayers.cpp:2338
jl_add_to_ee at /source/src/jitlayers.cpp:2338
jl_add_to_ee at /source/src/jitlayers.cpp:2338
jl_add_to_ee at /source/src/jitlayers.cpp:2338
[...]
```
Anyone got a clue? :grin:

gbaraldi commented:
@nanosoldier `runbenchmarks(ALL, vs=":master")`

Zentrik commented:
I think you missed the first line of the error
```
julia: /workspace/srcdir/llvm-project/llvm/include/llvm/Support/Casting.h:578: decltype(auto) llvm::cast(From*) [with To = llvm::IntegerType; From = llvm::Type]: Assertion `isa<To>(Val) && "cast<Ty>() argument of incompatible type!"' failed.
```

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/53070#issuecomment-2073010083) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/9ed8f43_vs_8945914/report.md).

gbaraldi commented:
Performance looks good!

gbaraldi commented:
All the crashes seem to be LoopVectorization/ @chriselrod verse packages, which I think are known to not work. Though we might still want to figure out why it causes a crash.

vchuravy commented:
[StrBase](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/1b5ee47_vs_903ffff/StrBase.primary.log)

Seems to hit the same error in IRCE

[TidierDB](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/1b5ee47_vs_903ffff/TidierDB.primary.log)
hits a sigabrt but no backtrace.

[StringAlgorithms](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/1b5ee47_vs_903ffff/StringAlgorithms.primary.log)
Fails the IR verifier

gbaraldi commented:
The string algorithms one is a know LLVM bug sigh

chriselrod commented:
I wouldn't worry about them too much

gbaraldi commented:
Unfortunately/Fortunately this seems to be an LLVM bug
```llvm
; ModuleID = 'containsnul'
source_filename = "containsnul"
target datalayout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-f80:128-n8:16:32:64-S128-ni:10:11:12:13"
target triple = "x86_64-unknown-linux-gnu"

@jl_nothing = external constant ptr
@jl_binding_ptr = private global ptr null

define swiftcc void @julia_containsnul_1(ptr nonnull swiftself %pgcstack_arg, ptr addrspace(10) noundef nonnull align 8 dereferenceable(16) %0) #0 !dbg !7 {
top:
  %pgcstack = call ptr @julia.get_pgcstack()
  %ptls_field = getelementptr inbounds ptr, ptr %pgcstack, i64 2
  %ptls_load = load ptr, ptr %ptls_field, align 8, !tbaa !11
  %1 = getelementptr inbounds ptr, ptr %ptls_load, i64 2
  %safepoint = load ptr, ptr %1, align 8, !tbaa !15, !invariant.load !10
  fence syncscope("singlethread") seq_cst
  call void @julia.safepoint(ptr %safepoint), !dbg !17
  fence syncscope("singlethread") seq_cst
  %2 = addrspacecast ptr addrspace(10) %0 to ptr addrspace(11), !dbg !18
  %3 = getelementptr inbounds i8, ptr addrspace(11) %2, i64 8, !dbg !18
  %4 = load ptr, ptr addrspace(11) %3, align 8, !dbg !18, !tbaa !19, !alias.scope !22, !noalias !25
  br label %L3, !dbg !18

L3:                                               ; preds = %found, %top
  %value_phi = phi ptr [ %4, %top ], [ %5, %found ]
  %5 = getelementptr i8, ptr %value_phi, i64 8, !dbg !30
  %.not = icmp ule ptr %5, %4, !dbg !34
  br i1 %.not, label %L13, label %L15, !dbg !33

L13:                                              ; preds = %L3
  %6 = load atomic ptr, ptr @jl_binding_ptr unordered, align 8, !dbg !40
  %.not2 = icmp eq ptr %6, null, !dbg !40
  br i1 %.not2, label %notfound, label %found, !dbg !40

L15:                                              ; preds = %L3
  ret void, !dbg !40

notfound:                                         ; preds = %L13
  %7 = call ptr @ijl_get_binding_or_error(ptr nonnull inttoptr (i64 140699009518448 to ptr), ptr nonnull inttoptr (i64 140699148750512 to ptr)), !dbg !40
  store atomic ptr %7, ptr @jl_binding_ptr release, align 8, !dbg !40
  br label %found, !dbg !40

found:                                            ; preds = %notfound, %L13
  %8 = phi ptr [ %6, %L13 ], [ %7, %notfound ], !dbg !40
  %9 = load atomic ptr addrspace(10), ptr %8 unordered, align 8, !dbg !40, !tbaa !41, !alias.scope !44, !noalias !45
  %.not3 = icmp eq ptr addrspace(10) %9, null, !dbg !40
  br i1 %.not3, label %err, label %L3, !dbg !40

err:                                              ; preds = %found
  call void @ijl_undefined_var_error(ptr addrspace(12) addrspacecast (ptr inttoptr (i64 140699148750512 to ptr) to ptr addrspace(12)), ptr addrspace(12) addrspacecast (ptr inttoptr (i64 140699009518448 to ptr) to ptr addrspace(12))), !dbg !40
  unreachable, !dbg !40
}

; Function Attrs: noinline optnone
define nonnull ptr addrspace(10) @jfptr_containsnul_2(ptr addrspace(10) %"function::Core.Function", ptr noalias nocapture noundef readonly %"args::Any[]", i32 %"nargs::UInt32") #1 {
top:
  %pgcstack = call ptr @julia.get_pgcstack()
  %0 = getelementptr inbounds ptr addrspace(10), ptr %"args::Any[]", i32 0
  %1 = load ptr addrspace(10), ptr %0, align 8, !tbaa !15, !invariant.load !10, !alias.scope !46, !noalias !47, !nonnull !10, !dereferenceable !48, !align !49
  call swiftcc void @julia_containsnul_1(ptr nonnull swiftself %pgcstack, ptr addrspace(10) %1)
  %2 = load ptr, ptr @jl_nothing, align 8, !tbaa !15, !invariant.load !10, !alias.scope !46, !noalias !47, !nonnull !10
  %3 = addrspacecast ptr %2 to ptr addrspace(10)
  ret ptr addrspace(10) %3
}

declare ptr @julia.get_pgcstack()

; Function Attrs: memory(argmem: readwrite, inaccessiblemem: readwrite)
declare void @julia.safepoint(ptr %0) #2

declare ptr @ijl_get_binding_or_error(ptr %0, ptr %1)

; Function Attrs: noreturn
declare void @ijl_undefined_var_error(ptr addrspace(12) %0, ptr addrspace(12) %1) #3

attributes #0 = { "frame-pointer"="all" "probe-stack"="inline-asm" }
attributes #1 = { noinline optnone "frame-pointer"="all" "probe-stack"="inline-asm" }
attributes #2 = { memory(argmem: readwrite, inaccessiblemem: readwrite) }
attributes #3 = { noreturn }

!llvm.module.flags = !{!0, !1, !2, !3, !4}
!llvm.dbg.cu = !{!5}

!0 = !{i32 2, !"Dwarf Version", i32 4}
!1 = !{i32 2, !"Debug Info Version", i32 3}
!2 = !{i32 2, !"julia.debug_level", i32 1}
!3 = !{i32 1, !"julia.__jit_debug_tsm_addr", i64 140728627667136}
!4 = !{i32 2, !"julia.optlevel", i32 2}
!5 = distinct !DICompileUnit(language: DW_LANG_Julia, file: !6, producer: "julia", isOptimized: true, runtimeVersion: 0, emissionKind: FullDebug, nameTableKind: GNU)
!6 = !DIFile(filename: "julia", directory: ".")
!7 = distinct !DISubprogram(name: "containsnul", linkageName: "julia_containsnul_1", scope: null, file: !8, line: 2, type: !9, scopeLine: 2, spFlags: DISPFlagDefinition | DISPFlagOptimized, unit: !5)
!8 = !DIFile(filename: "/home/gbaraldi/julia3/test.jl", directory: ".")
!9 = !DISubroutineType(types: !10)
!10 = !{}
!11 = !{!12, !12, i64 0}
!12 = !{!"jtbaa_gcframe", !13, i64 0}
!13 = !{!"jtbaa", !14, i64 0}
!14 = !{!"jtbaa"}
!15 = !{!16, !16, i64 0, i64 1}
!16 = !{!"jtbaa_const", !13, i64 0}
!17 = !DILocation(line: 2, scope: !7)
!18 = !DILocation(line: 3, scope: !7)
!19 = !{!20, !20, i64 0, i64 1}
!20 = !{!"jtbaa_memoryptr", !21, i64 0}
!21 = !{!"jtbaa_array", !13, i64 0}
!22 = !{!23}
!23 = !{!"jnoalias_typemd", !24}
!24 = !{!"jnoalias"}
!25 = !{!26, !27, !28, !29}
!26 = !{!"jnoalias_gcframe", !24}
!27 = !{!"jnoalias_stack", !24}
!28 = !{!"jnoalias_data", !24}
!29 = !{!"jnoalias_const", !24}
!30 = !DILocation(line: 316, scope: !31, inlinedAt: !33)
!31 = distinct !DISubprogram(name: "+;", linkageName: "+", scope: !32, file: !32, type: !9, spFlags: DISPFlagDefinition | DISPFlagOptimized, unit: !5)
!32 = !DIFile(filename: "pointer.jl", directory: ".")
!33 = !DILocation(line: 5, scope: !7)
!34 = !DILocation(line: 39, scope: !35, inlinedAt: !37)
!35 = distinct !DISubprogram(name: "|;", linkageName: "|", scope: !36, file: !36, type: !9, spFlags: DISPFlagDefinition | DISPFlagOptimized, unit: !5)
!36 = !DIFile(filename: "bool.jl", directory: ".")
!37 = !DILocation(line: 438, scope: !38, inlinedAt: !33)
!38 = distinct !DISubprogram(name: "<=;", linkageName: "<=", scope: !39, file: !39, type: !9, spFlags: DISPFlagDefinition | DISPFlagOptimized, unit: !5)
!39 = !DIFile(filename: "operators.jl", directory: ".")
!40 = !DILocation(line: 6, scope: !7)
!41 = !{!42, !42, i64 0}
!42 = !{!"jtbaa_binding", !43, i64 0}
!43 = !{!"jtbaa_data", !13, i64 0}
!44 = !{!28}
!45 = !{!26, !27, !23, !29}
!46 = !{!29}
!47 = !{!26, !27, !28, !23}
!48 = !{i64 16}
!49 = !{i64 8}
```
`./opt --passes=irce test.ll -S` to reproduce.

gbaraldi commented:
Opened issue upstream https://github.com/llvm/llvm-project/issues/89959

chriselrod commented:
@gbaraldi Is that a minimal reproducer from LoopVectorization.jl?
Maybe I'm one of the few people using pointers as loop induction variables, if the problem was only uncovered now that Julia master represents its pointers as `ptr` instead of `i$(8sizeof(Int))`.
In certain situations (depending on addressing mode, e.g. x86's), it can improve codegen and reduce the amount of instructions needed in certain circumstances. LV actually has a lot of code dedicated to identifying those circumstances.

gbaraldi commented:
That is from StrBase, not sure if the patch will fix the LV issue

gbaraldi committed:
Bump LLVM to 17.0.6+3

gbaraldi commented:
@nanosoldier `runtests(["StringAlgorithms", "TriangularSolve", "Pyehtim", "GPUCompiler", "SurveyDataWeighting", "TidierDB", "StrBase", "Jadex", "TropicalGEMM", "BenchmarkProfiles", "LabelledArrays", "RandomFeatures", "FinEtools", "PreallocationTools", "BLASBenchmarksCPU", "Ai4EComponentLib", "LibTrixi"])`

vchuravy commented on src/aotcompile.cpp:
```diff
@@ -1690,6 +1702,7 @@ void jl_dump_native_impl(void *native_code,
         JL_TIMING(NATIVE_AOT, NATIVE_Setup);
         dataM.setTargetTriple(TheTriple.str());
         dataM.setDataLayout(DL);
+        dataM.setPICLevel(PICLevel::BigPIC);
```
Why are we now requiring BigPIC?

gbaraldi replied:
https://github.com/JuliaLang/julia/pull/53070#issuecomment-2002972428 It's a weird SSP thing where they are getting it from the Module instead of the TM (IMO it's an issue with their API but oh well) and we use BigPIC because that's what the image gets linked as in the end anyway (-fPIC)

vchuravy commented on src/jitlayers.cpp:
```diff
@@ -1865,6 +1909,48 @@ Error JuliaOJIT::addObjectFile(orc::JITDylib &JD, std::unique_ptr<MemoryBuffer>
 }
 
+#if JL_LLVM_VERSION >= 170000
+Expected<ExecutorSymbolDef> JuliaOJIT::findSymbol(StringRef Name, bool ExportedSymbolsOnly)
```
Did only the return type change?

gbaraldi replied:
Pretty much, ExecutorSymbolDef is the new `JITEvaluatedSymbol`. It has a bit more type safety and supports authentication but the API is almost one to one wiht the exception of taking raw pointers which requires a `fromPtr` call

vchuravy commented on stdlib/LLD_jll/Project.toml:
```diff
@@ -1,6 +1,6 @@
 name = "LLD_jll"
 uuid = "d55e3150-da41-5e91-b323-ecfd1eec6109"
-version = "16.0.6+4"
+version = "17.0.6+3"
```
Is the patch version really the same?

gbaraldi replied:
Yes

chriselrod replied:
17.0.6 is the latest patch release with `LLVM_VERSION_MAJOR == 17`.

It seems 6 patch releases is the standard schedule:
https://llvm.org/docs/HowToReleaseLLVM.html#annual-release-schedule
Although with v18, they've switched to `LLVM_VERSION_MINOR == 1` instead of `0`.

giordano replied:
Actually yes: the branch https://github.com/JuliaLang/llvm-project/commits/julia-release/17.x/ builds on top of https://github.com/JuliaLang/llvm-project/commit/6009708b4367171ccdbf4b5905cb6a803753fe18, which is the tip of https://github.com/llvm/llvm-project/releases/tag/llvmorg-17.0.6

vchuravy commented on test/llvmpasses/llvmcall.jl:
```diff
@@ -1,11 +1,6 @@
-# RUN: export JULIA_LLVM_ARGS="--opaque-pointers=0"
-
-# RUN: julia --startup-file=no %s %t
-# RUN: cat %t/* | FileCheck %s --check-prefixes=CHECK,TYPED
```
We also need to drop lines prefixed with TYPED

nanosoldier commented:
The package evaluation job [you requested](https://github.com/JuliaLang/julia/pull/53070#issuecomment-2083039384) has completed - possible new issues were detected.
The [**full report**](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/b5a8598_vs_903ffff/report.html) is available.

gbaraldi committed:
Remove TYPED checks from llvmpasses

giordano committed:
Merge branch 'master' into bump-llvm17

gbaraldi commented:
@nanosoldier `runbenchmarks(ALL, vs=":master")`

gbaraldi commented:
@nanosoldier `runtests()`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/53070#issuecomment-2085423703) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/9353140_vs_831ebe0/report.md).

nanosoldier commented:
The package evaluation job [you requested](https://github.com/JuliaLang/julia/pull/53070#issuecomment-2085424491) has completed - possible new issues were detected.
The [**full report**](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/73ef4c6_vs_831ebe0/report.html) is available.
